### PR TITLE
Expose methods for skipping insertion point or removing speed bump

### PR DESCRIPTION
**Pull request by goldenapples**
Allows speed bumps to gracefully short-circuit their execution by
calling a static method on the Speed_Bumps class, like so:

```
return Speed_Bumps::return_false_and_skip()
```

(will ensure that nothing inserts at the current insertion point -
temporarily removing all other constraint checks so that they don't run
any potentially expensive operations - but reset the speed bump after
the current insertion point, so that it can still run over other checks
on the content.)

See #75.

### CODE REVIEW

**Comment by goldenapples** [comment]
This issue has continued to be a performance problem for us on posts with extremely long content. (We recently got flagged by VIP for performance issues on a 200-paragraph post which took over 6s to generate with all the speed bumps in it.)

This approach is similar to the route I took in #79, with the exception that instead of magic strings for return values, this uses static methods on the Speed_Bumps class. (instead of `return SPEED_BUMPS_RETURN_FALSE_AND_SKIP;`, here that would be `return Speed_Bumps::return_false_and_skip();`).

Either way, this allows for huge performance improvements in situations where a constraint knows that it wants to disallow any speed bump insertion at a given point - otherwise, every single constraint rule is check for every speed bump at each paragraph to determine whether a speed bump can be inserted at the current location.

I'm not sure I like this a whole lot better - it gets around the problems of using magic return values, but then it introduces functions with side effects, which I find a bit hard to read and reason about.

Maybe a better pattern to follow, which would still be possible with this code, would be for a filter to handle the logic in steps:

```
Speed_Bumps::skip_insertion_point( false );
return false;
```

as opposed to doing it in one line like:

```
return Speed_Bumps::return_false_and_skip()
```

I'd like some feedback on the code here overall, though, to see about getting to a point where it can be merged soon.

File: speed-bumps.php [hunk 1]
```diff
@@ -139,6 +141,8 @@ public function insert_speed_bumps( $the_content ) {
 						'inserted_content' => $content_to_be_inserted,
 					);
 				}
+
+				do_action( 'done_speed_bump_constraints', $speed_bump_filter );
```
Comment (danielbachhuber) [inline speed-bumps.php]: This might be better as `speed_bumps_constraints_completed`

File: speed-bumps.php [hunk 2]
```diff
@@ -262,11 +266,103 @@ public function clear_all_speed_bumps() {
...
+	 *
+	 * @param string $speed_bump_id Defaults to the value of current_filter().
+	 */
+	public static function remove_speed_bump( $speed_bump_id = null ) {
+		global $_wp_filters_backed_up, $wp_filter;
+
+		if ( $speed_bump_id ) {
+			$filter_id = sprintf( self::$filter_id, $speed_bump_id );
+		} else {
+			$filter_id = current_filter();
```
Comment (danielbachhuber) [inline speed-bumps.php]: I'm not a huge fan of methods that are conditionally dependent on globals because it makes it easy to introduce a potentially nasty bug if you accidentally forget to provide the speed bump id.

Instead, the speed bump id should always be provided, even if it means calling the method with `remove_speed_bump( current_filter() )`

**Comment by danielbachhuber** [comment]
> This approach is similar to the route I took in #79, with the exception that instead of magic strings for return values, this uses static methods on the Speed_Bumps class.
>
> [...]
>
> I'm not sure I like this a whole lot better - it gets around the problems of using magic return values, but then it introduces functions with side effects, which I find a bit hard to read and reason about.

I do like this better conceptually, although I agree with your sentiment.

What about introducing a `status` concept for the Speed Bumps class, where you're always just changing the value of the status attribute?

> I'd like some feedback on the code here overall, though, to see about getting to a point where it can be merged soon.

Anything else in particular that you're looking for feedback on?

File: speed-bumps.php (hunk 2 above)
Comment (goldenapples) [inline speed-bumps.php]: That's a good point, although I would prefer a method that always removed the current speed bump, since that's the only use case I was concerned with enabling here. (I only added the optional `$speed_bump_id` parameter there for tests.) 

I'm checking that the current_filter value is one of the speed bumps registered through this plugin [here](https://github.com/fusioneng/speed-bumps/blob/f88348a65e02d2472e7c699bab027cb91e8bec51/speed-bumps.php#L306). I can just remove the optional `$speed_bump_id` parameter to avoid unclearness.

File: speed-bumps.php (hunk 1 above)
Comment (goldenapples) [inline speed-bumps.php]: :+1: Good call, I'll do a pass over the plugin for naming consistancy.

**Comment by goldenapples** [comment]
@danielbachhuber Thanks for looking this over for me!

> What about introducing a status concept for the Speed Bumps class, where you're always just changing the value of the status attribute?

I think that given the sheer amount of possibilities for status on the Speed Bumps object as a whole, that seems as though it would get unwieldy. In the most basic configuration, there could be any number of speed bumps, times any number of rules on each, times each possible insertion point. And that doesn't even touch cases where most than one piece of content will be processed in a given pageload.

Given the current architecture which is based heavily on WP's core filter hooks, I guess a flow involving removing filters and re-adding them was the best way of dealing with performance issues I could come up with. 

> Anything else in particular that you're looking for feedback on?

Basically just the general approach - does it make sense to remove filters on a hook in the middle of processing that hook? Is there a case where that will have unintended consequences? I couldn't find a case where that wouldn't work: since the looping inside `apply_filters` is using a `next()` iterator, modifying the array being iterated over midway through works just fine. But I just wanted a second opinion as to whether this was dangerous to try.

**Comment by danielbachhuber** [comment]
> I think that given the sheer amount of possibilities for status on the Speed Bumps object as a whole, that seems as though it would get unwieldy.

Fair enough.

> Given the current architecture which is based heavily on WP's core filter hooks, I guess a flow involving removing filters and re-adding them was the best way of dealing with performance issues I could come up with.

Yes, this makes sense to me.

> Basically just the general approach - does it make sense to remove filters on a hook in the middle of processing that hook? Is there a case where that will have unintended consequences?

It would take me an hour of diving into the code to think through whether there might be any unintended consequences. Fortunately, we have great test coverage! Given our currently level of test coverage, I feel (rightfully or not) pretty comfortable with the logic of the changes.

**Comment by goldenapples** [comment]
I think this is cleaned up enough to be ready to merge. I've simplified all the static methods on the Speed_Bumps class to have a single responsibility each - rather than have methods which can either take an argument or fall back to a global, I have two separate methods (`remove_current_speed_bump`, which always uses the global filter id, and `remove_speed_bump`, which always takes an argument). Also, the function which remove or skip speed bumps now do not return anything. Skipping an insertion point and returning false is done through helper methods like `return_false_and_skip`.

**Comment by ryanboswell** [comment]
I like the cleanup you did with the separation of methods here, the whole thing is a lot easier to follow at a glance. I think this is in a good place to 🚢
